app/auth/views/proton.py

```diff
@@ -0,0 +1,108 @@
+from flask import request, session, redirect, flash, url_for
+from flask_limiter.util import get_remote_address
+from flask_login import current_user
+from requests_oauthlib import OAuth2Session
+
+from app.auth.base import auth_bp
+from app.auth.views.login_utils import after_login
+from app.config import (
+    PROTON_BASE_URL,
+    PROTON_CLIENT_ID,
+    PROTON_CLIENT_SECRET,
+    PROTON_VALIDATE_CERTS,
+    URL,
+)
+from app.proton.proton_client import HttpProtonClient, convert_access_token
+from app.proton.proton_callback_handler import ProtonCallbackHandler, Action
+from app.utils import encode_url, sanitize_next_url
+
+_authorization_base_url = PROTON_BASE_URL + "/oauth/authorize"
+_token_url = PROTON_BASE_URL + "/oauth/token"
+
+# need to set explicitly redirect_uri instead of leaving the lib to pre-fill redirect_uri
+# when served behind nginx, the redirect_uri is localhost... and not the real url
+_redirect_uri = URL + "/auth/proton/callback"
+
+
+def extract_action() -> Action:
+    action = request.args.get("action")
+    if action is not None:
+        if action == "link":
+            return Action.Link
+        else:
+            raise Exception(f"Unknown action: {action}")
+    return Action.Login
+
+
+def get_action_from_state() -> Action:
+    oauth_action = session["oauth_action"]
+    if oauth_action == Action.Login.value:
+        return Action.Login
+    elif oauth_action == Action.Link.value:
+        return Action.Link
+    raise Exception(f"Unknown action in state: {oauth_action}")
+
+
+@auth_bp.route("/proton/login")
+def proton_login():
+    if PROTON_CLIENT_ID is None or PROTON_CLIENT_SECRET is None:
+        return redirect(url_for("auth.login"))
+
+    next_url = sanitize_next_url(request.args.get("next"))
+    if next_url:
+        redirect_uri = _redirect_uri + "?next=" + encode_url(next_url)
+    else:
+        redirect_uri = _redirect_uri
+
+    proton = OAuth2Session(PROTON_CLIENT_ID, redirect_uri=redirect_uri)
+    authorization_url, state = proton.authorization_url(_authorization_base_url)
+
+    # State is used to prevent CSRF, keep this for later.
+    session["oauth_state"] = state
+    session["oauth_action"] = extract_action().value
+    return redirect(authorization_url)
+
+
+@auth_bp.route("/proton/callback")
+def proton_callback():
```

Comment:
As I understand, this callback is used for both logging user and linking user right? In this case, maybe the code can be easier to read if we have something like:

```
if current_user:
     handle_link_user()
else:
     handle_login()
```

what do you think?

Reply:
The initial logic for handling the request is the same in both cases, until we call the `ProtonCallbackHandler`.

The code inside the `ProtonCallbackHandler` already determines which case we're dealing with (and performs the `if` by checking also the `action` that we have stored in the state):
https://github.com/cquintana92/app/blob/feature/add-login-with-proton/app/proton/proton_callback_handler.py#L209

Reply:
Yeah I saw that too, I just think once we have proton user data, it's maybe easier to read if we separate these 2 cases in the endpoint code directly (and not inside the ProtonCallbackHandler.handle -> process_case which adds additional call stack). This is totally optional.

Reply:
I agree, the sooner we separate the logic flows the easier is to read.

Reply:
Pushed the changes to separate the cases on the controller level